aptos-move/framework/aptos-framework/sources/nonce_validation.move

```diff
@@ -0,0 +1,148 @@
+        // is last added to the nonces[i] map.
+        last_stored_times: vector<u64>,
+    }
+
+    struct NonceKey has copy, drop, store {
+        sender_address: address,
+        nonce: u64,
+    }
+
+    struct NonceHistory has key {
+        // Key = sip_hash(NonceKey) % NUM_BUCKETS
+        // Value = Bucket
+        nonce_table: Table<u64, Bucket>,
+        // Used to facilitate prefill the nonce_table with empty buckets
+        // one by one using `add_nonce_bucket` method.
+        // This is the next_key to prefill with an empty bucket
+        next_key: u64,
+    }
+
+    public(friend) fun initialize(aptos_framework: &signer) {
+        initialize_nonce_table(aptos_framework);
+    }
+
+    public entry fun initialize_nonce_table(aptos_framework: &signer) {
+        if (!exists<NonceHistory>(@aptos_framework)) {
+            let table = table::new();
+            let nonce_history = NonceHistory {
+                nonce_table: table,
+                next_key: 0,
+            };
+            move_to<NonceHistory>(aptos_framework, nonce_history);
+        };
+    }
+
+    // This method is used to prefill the nonce_table with empty buckets one by one.
+    public entry fun add_nonce_bucket(aptos_framework: &signer) acquires NonceHistory {
+        if (exists<NonceHistory>(@aptos_framework)) {
+            let current_time = timestamp::now_seconds();
+            let nonce_history = borrow_global_mut<NonceHistory>(@aptos_framework);
+            if (!table::contains(&nonce_history.nonce_table, nonce_history.next_key)) {
+                let nonces = vector::empty();
+                let last_stored_times = vector::empty();
+                vector::push_back(&mut nonces, big_ordered_map::new());
+                vector::push_back(&mut nonces, big_ordered_map::new());
+                vector::push_back(&mut last_stored_times, 0);
+                vector::push_back(&mut last_stored_times, 0);
+                table::add(&mut nonce_history.nonce_table, nonce_history.next_key, Bucket {
+                    nonces: nonces,
+                    last_stored_times: last_stored_times,
+                });
+            };
+            nonce_history.next_key = nonce_history.next_key + 1;
+        };
+    }
+
+    // Returns true if the input (address, nonce) pair doesn't exist in the nonce history, and inserted into nonce history successfully.
+    // Returns false if the input (address, nonce) pair already exists in the nonce history.
+    public(friend) fun check_and_insert_nonce(
+        sender_address: address,
+        nonce: u64,
+        txn_expiration_time: u64,
+    ): bool acquires NonceHistory {
+        let nonce_history = borrow_global_mut<NonceHistory>(@aptos_framework);
+        let nonce_key = NonceKey {
+            sender_address,
+            nonce,
+        };
+        let index = sip_hash_from_value(&nonce_key) % NUM_BUCKETS;
+        let map_index = (txn_expiration_time / MAP_SWITCH_TIME) % 2;
```

Comment:
I don't quite get what's this and how is it used?

Reply:
We are using 2 big ordered maps per bucket to make garbage collection easy.
At any time, only of the maps is considered "active".
We search for the given input (address, nonce) pair in both the big ordered maps.
If the (address, nonce) pair doesn't exist, we insert the entry into only the active map.
Here, I'm switching which of the big ordered map is considered "active" once every 75 seconds.
Does it make sense?